### PR TITLE
Add BackgroundTransformer which decodes responses in an isolate

### PR DESCRIPTION
**Pull request by ueman**
Fixes https://github.com/cfug/diox/issues/43

This PR vendors the [`compute()`](https://api.flutter.dev/flutter/foundation/compute-constant.html) related code from Flutter. After this library requires Dart 3, we can remove that code and just use `Isolate.run()`. By vendoring the Flutter compute code, we avoid having a dependency on Flutter and we're still compatible with web and io platforms.

This code also uses the trick from [`AssetBundle.loadString()`](https://api.flutter.dev/flutter/services/AssetBundle/loadString.html) to only offload huge json responses.

Open points for discussion:
- [x] Should this become the default transformer?
- [x] Should I also vendor test code for compute related code?
- [x] Should we remove `diox_flutter_transformer` from the list of diox plugins, since we provide this out of the box?
- [ ] ~Is there a way for json encoding to be offloaded to a background isolate?~ At least there's no easy way

Open tasks:
- [x] Add tests for background transformer

### New Pull Request Checklist

- [x] I have read the [Documentation](https://pub.dev/documentation/diox/latest/)
- [x] I have searched for a similar pull request in the [project](https://github.com/cfug/diox/pulls) and found none
- [x] I have updated this branch with the latest `main` branch to avoid conflicts (via merge from master or rebase)
- [x] I have added the required tests to prove the fix/feature I'm adding
- [ ] I have updated the documentation (if necessary)
- [x] I have run the tests without failures

### CODE REVIEW

File: dio/lib/src/transformers/background_transformer/compute.dart [hunk 1]
```diff
@@ -0,0 +1,134 @@
+/// This file corresponds to Flutter's
```
Comment (AlexV525) [inline dio/lib/src/transformers/background_transformer/compute.dart:1]: Can we raise the level to `src/compute/*`?

File: dio/lib/src/transformers/background_transformer/background_transformer.dart [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import 'dart:async';
+import 'dart:convert';
+import 'package:diox/diox.dart';
+import 'compute.dart';
+
+/// [BackgroundTransformer] will do the deserialization of JSON
+/// in a background isolate if possible.
+class BackgroundTransformer extends DefaultTransformer {
```
Comment (AlexV525) [inline dio/lib/src/transformers/background_transformer/background_transformer.dart:8]: A test should be added correspondingly.

**Comment by AlexV525** [comment]
> Should this become the default transformer?

I would vote yes. 👍

> Should I also vendor test code for compute related code?

Seems unnecessary at present.

> Should we remove `diox_flutter_transformer` from the list of diox plugins, since we provide this out of the box?

I think it's okay, but let's consult @arklabsnz.

> Is there a way for json encoding to be offloaded to a background isolate?

I think the sending JSON would be small in most cases, or we can to a similar thing as the decoding.

File: dio/lib/src/transformers/background_transformer/compute.dart (hunk 1 above)
Comment (ueman) [reply dio/lib/src/transformers/background_transformer/compute.dart:1]: Sure

**Comment by ueman** [comment]
> I think the sending JSON would be small in most cases, or we can to a similar thing as the decoding.

The problem is that there's no easy way to figure out the size of whatever is getting send and there's no way of being sure that everything is valid to send to another isolate.

So I would keep that sync right now.

File: dio/lib/src/transformers/background_transformer/compute.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (ueman) [reply dio/lib/src/transformers/background_transformer/compute.dart:1]: I'll change it

File: dio/lib/src/transformer.dart [hunk 3]
```diff
@@ -63,19 +63,24 @@ abstract class Transformer {
...
 
 class DefaultTransformer extends Transformer {
-  DefaultTransformer({this.jsonDecodeCallback});
+  DefaultTransformer({
+    this.jsonDecodeCallback = jsonDecode,
+    this.jsonEncodeCallback = jsonEncode,
+  });
 
-  JsonDecodeCallback? jsonDecodeCallback;
+  JsonDecodeCallback jsonDecodeCallback;
```
[the remark this answers is not in the document]
Comment (ueman) [reply dio/lib/src/transformer.dart:76]: I'm not sure I understand what you want me to do. Can you explain it a bit more?

[the remark this answers is not in the document]
Comment (AlexV525) [reply dio/lib/src/transformer.dart:76]: > I'm not sure I understand what you want me to do. Can you explain it a bit more?

Nevermind, I think the change is good enough.

File: dio/CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,9 @@
 # CHANGELOG
 
+## UNRELEASED
+
+- Change `Dio.transformer` from `DefaultTransformer` to `BackgroundTransformer`
```
Comment (AlexV525) [inline dio/CHANGELOG.md:5]:
```suggestion
- Change `Dio.transformer` from `DefaultTransformer` to `BackgroundTransformer`.
```

File: dio/lib/diox.dart [hunk 5]
```diff
@@ -17,3 +17,4 @@ export 'src/parameter.dart';
 export 'src/redirect_record.dart';
 export 'src/response.dart';
 export 'src/transformer.dart';
+export 'src/transformers/background_transformer/background_transformer.dart';
```
Comment (AlexV525) [inline dio/lib/diox.dart:20]: Any specific reason that we need to have `/background_transformer` directory here?

Comment (ueman) [reply dio/lib/diox.dart:20]: No, previously the compute function was also in this folder. When moving it, I must have forgotten about this folder.

**Comment by josh-burton** [comment]
I'm happy to deprecate diox_flutter_transformer if the functionality is built in :)

**Review by AlexV525 — APPROVED** [review]
LGTM with some other nits

File: dio/lib/src/transformer.dart [hunk 6]
```diff
@@ -57,25 +56,32 @@ abstract class Transformer {
...
 ///
 /// If you want to custom the transformation of request/response data,
 /// you can provide a [Transformer] by your self, and replace
 /// the [DefaultTransformer] by setting the [dio.transformer].
-
-typedef JsonDecodeCallback = dynamic Function(String);
-
 class DefaultTransformer extends Transformer {
-  DefaultTransformer({this.jsonDecodeCallback});
+  DefaultTransformer({
```
Comment (AlexV525) [inline dio/lib/src/transformer.dart:71]: Regarding changing the default transformer to the new background one, can we deprecate this by type alias, maybe `SynchronizedTransformer` or `ForegroundTransformer` (v.s. `BackgroundTransformer`)?
```dart
@Deprecated('Use BackgroundTransformer instead')
typedef DefaultTransformer = SynchronizedTransformer;
```

Comment (ueman) [reply dio/lib/src/transformer.dart:71]: Sure, will do 👍

Comment (ueman) [reply dio/lib/src/transformer.dart]: Done. I've also moved the old `DefaultTransformer` into the transformer folder.

File: dio/lib/src/transformers/sync_transformer.dart [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+/// The callback definition for encoding a JSON object.
+typedef JsonEncodeCallback = FutureOr<String> Function(Object);
+
+/// The default [Transformer] for [Dio].
+///
+/// If you want to custom the transformation of request/response data,
+/// you can provide a [Transformer] by your self, and replace
+/// the [DefaultTransformer] by setting the [dio.transformer].
+class SyncTransformer extends Transformer {
```
Comment (AlexV525) [inline dio/lib/src/transformers/sync_transformer.dart:24]: This should also be deprecated.

**Review by AlexV525 — APPROVED** [review]

File: dio/lib/src/transformers/sync_transformer.dart (hunk 7 above)
Comment (ueman) [reply dio/lib/src/transformers/sync_transformer.dart:24]: I don't think so. There are valid use cases for this, like web apps (they're single threaded), server applications (where you can't spawn isolates for every request).

Renaming and deprecating the old name made a lot of sense though.

**Comment by kuhnroyal** [comment]
The `Isolate.exit` function does not exist in the min Dart SDK version.

**Comment by kuhnroyal** [comment]
Was introduced in 2.15: https://github.com/dart-lang/sdk/commit/56403a0d356a3508b043f13e7f7168df3003d529

**Pull request merged**